tools/gfx/render.h

```diff
@@ -315,33 +311,53 @@ class BufferResource: public Resource
             elementSize = 0;
             format = Format::Unknown;
         }
-            /// Set up default parameters based on usage
-        void setDefaults(Usage initialUsage);
-
+        void setDefaults(Usage initialUsage)
+        {
+            if (bindFlags == 0)
+            {
+                bindFlags = getDefaultBindFlagsFromUsage(initialUsage);
+            }
+        }
         size_t sizeInBytes;     ///< Total size in bytes
         int elementSize;        ///< Get the element stride. If > 0, this is a structured buffer
         Format format;
     };
-
-        /// Get the buffer description
-    SLANG_FORCE_INLINE const Desc& getDesc() const { return m_desc; }
-
-        /// Ctor
-    BufferResource(const Desc& desc):
-        Parent(Type::Buffer),
-        m_desc(desc)
-    {
+    virtual SLANG_NO_THROW Desc* SLANG_MCALL getDesc() = 0;
+};
+#define SLANG_UUID_IBufferResource                                                     \
+    {                                                                                  \
+        0x1b274efe, 0x5e37, 0x492b, { 0x82, 0x6e, 0x7e, 0xe7, 0xe8, 0xf5, 0xa4, 0x9b } \
     }
 
-    protected:
-    Desc m_desc;
-};
+template <typename T> T _slang_gfx_max(T v0, T v1) { return v0 > v1 ? v0 : v1; }
+
+static inline unsigned int _slang_gfx_ones32(unsigned int x)
```

Comment:
Having this stuff in the header sure comes across like a code smell. It is a sign that we have too much logic in the `Desc` types.

Hypothetically, we could eventually split things so that there is `gfx.h` which is the minimal header that just defines the COM interface-y stuf, and then `gfxu.h/.cpp` which is "user"/utility code that is meant to be directly added to your project (similar to D3DX, but distributed as source).